File: src/tidal.py
#!/usr/bin/env python3
"""
tidal.py — Subproceso intermediario entre Rust y Tidal via tidalapi.

Uso:
  python3 tidal.py auth start
  python3 tidal.py auth poll
  python3 tidal.py search "Daft Punk"
  python3 tidal.py stream 12345
  python3 tidal.py stream 12345 HI_RES_LOSSLESS

Siempre imprime JSON en stdout. Errores van a stderr.
Instalar: pip install tidalapi --break-system-packages
"""

import json
import sys
import threading
from pathlib import Path

import tidalapi

SESSION_FILE = str(Path.home() / ".config" / "tidal-tui" / "tidalapi_session.json")
POLL_FILE    = str(Path.home() / ".config" / "tidal-tui" / "oauth_pending.json")

def out(data: dict):
    print(json.dumps(data, ensure_ascii=False))
    sys.stdout.flush()

def err(msg: str):
    print(msg, file=sys.stderr)
    out({"error": msg})

def make_session() -> tidalapi.Session:
    return tidalapi.Session()

def load_session(session: tidalapi.Session) -> bool:
    try:
        session.load_oauth_session(SESSION_FILE)
        return session.check_login()
    except Exception:
        return False

def save_session(session: tidalapi.Session):
    Path(SESSION_FILE).parent.mkdir(parents=True, exist_ok=True)
    session.save_oauth_session(SESSION_FILE)

# ─── Comandos ─────────────────────────────────────────────────────────────────

def cmd_auth_start():
    """Inicia OAuth Device Flow. Guarda estado en POLL_FILE para que poll lo use."""
    session = make_session()

    # login_oauth() devuelve (future, url_obj)
    # Necesitamos guardar el future en un hilo y el url para mostrarlo al usuario
    result = {}
    event  = threading.Event()

    def do_login():
        try:
            future, url = session.login_oauth()
            result["url"]  = str(url)
            result["code"] = getattr(url, "user_code", "")
            event.set()
            future.result()  # bloquea hasta que el usuario autorice
            save_session(session)
            Path(POLL_FILE).write_text(json.dumps({"done": True}))
        except Exception as e:
            Path(POLL_FILE).write_text(json.dumps({"done": False, "error": str(e)}))

    t = threading.Thread(target=do_login, daemon=True)
    t.start()
    event.wait(timeout=10)

    if "url" not in result:
        err("Timeout esperando URL de auth")
        return

    # Guardar el hilo en POLL_FILE como "pendiente"
    Path(POLL_FILE).write_text(json.dumps({"done": False}))

    out({"url": result["url"], "code": result["code"]})

    # Esperar en background a que el usuario autorice
    t.join()

def cmd_auth_poll():
    """Verifica si el OAuth ya completó (lee POLL_FILE)."""
    poll_path = Path(POLL_FILE)
    session   = make_session()

    if load_session(session):
        out({"authenticated": True})
        return

    if not poll_path.exists():
        out({"authenticated": False, "pending": False})
        return

    try:
        state = json.loads(poll_path.read_text())
        if state.get("done"):
            poll_path.unlink(missing_ok=True)
            out({"authenticated": True})
        else:
            out({"authenticated": False, "pending": True, "error": state.get("error", "")})
    except Exception as e:
        out({"authenticated": False, "error": str(e)})

def cmd_search(query: str, limit: int = 20):
    session = make_session()
    if not load_session(session):
        err("No autenticado")
        return

    try:
        results = session.search(query, [tidalapi.Track], limit=limit)
        tracks  = results.get("tracks", []) or []
        out([_track_dict(t) for t in tracks[:limit]])
    except Exception as e:
        err(str(e))

def cmd_stream(track_id: int, quality_str: str = "LOSSLESS"):
    session = make_session()
    if not load_session(session):
        err("No autenticado")
        return

    quality_map = {
        "HI_RES_LOSSLESS": tidalapi.Quality.hi_res_lossless,
        "LOSSLESS":        tidalapi.Quality.lossless,
        "HIGH":            tidalapi.Quality.high,
    }
    session.config.quality = quality_map.get(quality_str, tidalapi.Quality.lossless)

    try:
        track = session.track(track_id)
        url   = track.get_url()
        out({
            "url":         url,
            "codec":       "flac",
            "bit_depth":   24 if quality_str == "HI_RES_LOSSLESS" else 16,
            "sample_rate": 96000 if quality_str == "HI_RES_LOSSLESS" else 44100,
            "mime_type":   "audio/flac",
        })
    except Exception as e:
        err(str(e))

def cmd_cover(track_id: int):
    session = make_session()
    if not load_session(session):
        err("No autenticado")
        return
    try:
        track = session.track(track_id)
        url   = track.album.image(640)
        out({"url": url, "title": track.name, "artist": track.artists[0].name, "album": track.album.name})
    except Exception as e:
        err(str(e))

def cmd_track_radio(track_id: int):
    session = make_session()
    if not load_session(session):
        err("No autenticado")
        return
    try:
        track = session.track(track_id)
        tracks = track.get_track_radio()
        out([_track_dict(t) for t in tracks])
    except Exception as e:
        err(str(e))

# ─── Helpers ──────────────────────────────────────────────────────────────────

def _track_dict(t: tidalapi.Track) -> dict:
    return {
        "id":            t.id,
        "title":         t.name,
        "duration":      t.duration,
        "track_number":  getattr(t, "track_num", None),
        "artists":       [{"id": a.id, "name": a.name} for a in t.artists],
        "album":         {"id": t.album.id, "title": t.album.name},
        "audio_quality": str(getattr(t, "audio_quality", "") or ""),
        "explicit":      getattr(t, "explicit", False),
    }

# ─── Entry point ──────────────────────────────────────────────────────────────

if __name__ == "__main__":
    args = sys.argv[1:]

    if not args:
        err("Uso: tidal.py <auth start|auth poll|search <query>|stream <id> [quality]|radio <id>>")
        sys.exit(1)

    match args:
        case ["auth", "start"]:
            cmd_auth_start()
        case ["auth", "poll"]:
            cmd_auth_poll()
        case ["search", query]:
            cmd_search(query)
        case ["search", query, limit]:
            cmd_search(query, int(limit))
        case ["stream", track_id]:
            cmd_stream(int(track_id))
        case ["stream", track_id, quality]:
            cmd_stream(int(track_id), quality)
        case ["radio", track_id]:
            cmd_track_radio(int(track_id))
        case ["cover", track_id]:
            cmd_cover(int(track_id))
        case _:
            err(f"Comando desconocido: {args}")
            sys.exit(1)

File: src/tidal.rs
/// tidal.rs — Llama a tidal.py como subproceso y parsea el JSON que devuelve.

use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};
use std::io::BufRead;
use std::process::{Command, Stdio};

// ─── Calidades ────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Quality {
    HiResLossless,
    Lossless,
    High,
}

impl Quality {
    pub fn as_api_str(&self) -> &'static str {
        match self {
            Quality::HiResLossless => "HI_RES_LOSSLESS",
            Quality::Lossless      => "LOSSLESS",
            Quality::High          => "HIGH",
        }
    }
    pub fn label(&self) -> &'static str {
        match self {
            Quality::HiResLossless => "HiRes FLAC 24bit",
            Quality::Lossless      => "FLAC 16bit/44.1kHz",
            Quality::High          => "AAC 320kbps",
        }
    }
}

// ─── Modelos ──────────────────────────────────────────────────────────────────

#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Artist {
    pub id:   u64,
    pub name: String,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Album {
    pub id:    u64,
    pub title: String,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Track {
    pub id:            u64,
    pub title:         String,
    pub duration:      u64,
    pub track_number:  Option<u32>,
    pub artists:       Vec<Artist>,
    pub album:         Album,
    pub audio_quality: Option<String>,
    pub explicit:      Option<bool>,
}

impl Track {
    pub fn artist_names(&self) -> String {
        self.artists.iter().map(|a| a.name.as_str()).collect::<Vec<_>>().join(", ")
    }

    pub fn duration_str(&self) -> String {
        let m = self.duration / 60;
        let s = self.duration % 60;
        format!("{m}:{s:02}")
    }

    pub fn quality_icon(&self) -> &'static str {
        match self.audio_quality.as_deref() {
            Some("HI_RES_LOSSLESS") => "◈",
            Some("LOSSLESS")        => "◆",
            _                       => "◇",
        }
    }
}

// Modelo para álbumes de la colección
#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FavAlbum {
    pub id:               u64,
    pub title:            String,
    pub number_of_tracks: u32,
    pub duration:         u64,
    pub artists:          Vec<Artist>,
    pub cover_url:        Option<String>,
}

impl FavAlbum {
    pub fn artist_names(&self) -> String {
        self.artists.iter().map(|a| a.name.as_str()).collect::<Vec<_>>().join(", ")
    }
}

#[derive(Debug, Clone)]
pub struct StreamInfo {
    pub url:         String,
    pub bit_depth:   u32,
    pub sample_rate: u32,
    pub codec:       String,
}

#[derive(Debug, Clone)]
pub struct CoverInfo {
    pub url: String,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Playlist {
    pub uuid:             String,
    pub title:            String,
    pub number_of_tracks: u32,
    pub duration:         u64,
    #[serde(rename = "type")]
    pub playlist_type:    String,   // "USER", "EDITORIAL", "ARTIST"
    pub public_playlist:  Option<bool>,
    pub image:            Option<String>,
    pub square_image:     Option<String>,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Mix {
    pub id:    String,
    pub title: String,
    pub sub_title: Option<String>,
}

// ─── Respuestas internas ──────────────────────────────────────────────────────

#[derive(Deserialize)]
struct AuthStartResp {
    url:   Option<String>,
    code:  Option<String>,
    error: Option<String>,
}

#[derive(Deserialize)]
struct AuthPollResp {
    authenticated: Option<bool>,
    #[serde(default)]
    error: String,
}

#[derive(Deserialize)]
struct StreamResp {
    url:         Option<String>,
    codec:       Option<String>,
    bit_depth:   Option<u32>,
    sample_rate: Option<u32>,
    error:       Option<String>,
}

#[derive(Deserialize)]
struct CoverResp {
    url:   Option<String>,
    error: Option<String>,
}

// ─── Cliente ──────────────────────────────────────────────────────────────────

pub struct TidalClient {
    pub quality:     Quality,
    pub script_path: String,
    pub python_path: String,
}

impl TidalClient {
    pub fn new() -> Self {
        let script_path = std::env::current_exe()
            .ok()
            .and_then(|p| p.parent().map(|d| d.join("tidal.py")))
            .filter(|p| p.exists())
            .or_else(|| std::env::current_dir().ok().map(|d| d.join("tidal.py")))
            .filter(|p| p.exists())
            .map(|p| p.to_string_lossy().to_string())
            .unwrap_or_else(|| "tidal.py".to_string());

        let python_path = std::env::var("TUIDAL_PYTHON_PATH").unwrap_or_else(|_| "python3".to_string());

        Self { quality: Quality::Lossless, script_path, python_path }
    }

    pub fn with_path_and_quality(script_path: String, quality: Quality, python_path: String) -> Self {
        Self { quality, script_path, python_path }
    }

    fn run(&self, args: &[&str]) -> Result<String> {
        // println!("[tidal.rs] script: {} args: {:?}", self.script_path, args);
        let output = Command::new(&self.python_path)
            .arg(&self.script_path)
            .args(args)
            .output()
            .map_err(|e| anyhow!(
                "No se pudo ejecutar python3: {e}\n¿Está tidal.py en '{}'?",
                self.script_path
            ))?;

        if !output.stderr.is_empty() {
            eprintln!("[tidal.py] {}", String::from_utf8_lossy(&output.stderr).trim());
        }

        let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
        if stdout.is_empty() {
            return Err(anyhow!("tidal.py no produjo output"));
        }
        Ok(stdout)
    }

   pub async fn get_favorite_tracks(&self) -> Result<Vec<Track>> {
        let stdout = self.run(&["fav_tracks"])?;
        let tracks: Vec<Track> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(tracks)
    }

    pub async fn get_favorite_albums(&self) -> Result<Vec<FavAlbum>> {
        let stdout = self.run(&["fav_albums"])?;
        let albums: Vec<FavAlbum> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(albums)
    }

    pub async fn get_album_tracks(&self, album_id: u64) -> Result<Vec<Track>> {
        let stdout = self.run(&["album_tracks", &album_id.to_string()])?;
        let tracks: Vec<Track> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(tracks)
    }

    // ── Auth ──────────────────────────────────────────────────────────────────

    pub async fn load_session(&self) -> Result<()> {
        let stdout = self.run(&["auth", "poll"])?;
        let resp: AuthPollResp = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        if resp.authenticated.unwrap_or(false) { Ok(()) }
        else { Err(anyhow!("Sin sesión activa")) }
    }

    pub async fn start_device_auth(&self) -> Result<(String, String, String)> {
        let script_path = self.script_path.clone();
        let python_path = self.python_path.clone();
        let stdout = tokio::task::spawn_blocking(move || {
            let mut child = Command::new(&python_path)
                .arg(&script_path)
                .args(["auth", "start"])
                .stdin(Stdio::null())
                .stderr(Stdio::null())
                .stdout(Stdio::piped())
                .spawn()
                .map_err(|e| anyhow!("Error lanzando python3: {e}"))?;
            // Leer solo la primera línea (JSON con URL), luego dejar el proceso
            // corriendo en background — el thread de Python sigue esperando la
            // autorización y escribe el resultado en POLL_FILE.
            let reader = std::io::BufReader::new(child.stdout.take().unwrap());
            let first_line = reader.lines()
                .next()
                .ok_or_else(|| anyhow!("tidal.py no produjo output"))?
                .map_err(|e| anyhow!("Error leyendo output: {e}"))?;
            Ok::<String, anyhow::Error>(first_line)
        }).await??;

        let resp: AuthStartResp = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        if let Some(e) = resp.error { return Err(anyhow!("{e}")); }
        let url  = resp.url.ok_or_else(|| anyhow!("Sin URL de auth"))?;
        let code = resp.code.unwrap_or_default();
        Ok(("pending".to_string(), code, url))
    }

    pub async fn poll_device_token(&self, _device_code: &str) -> Result<bool> {
        let stdout = self.run(&["auth", "poll"])?;
        let resp: AuthPollResp = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        if !resp.error.is_empty() { eprintln!("[auth poll] {}", resp.error); }
        Ok(resp.authenticated.unwrap_or(false))
    }

    // ── API ───────────────────────────────────────────────────────────────────

    pub async fn search(&self, query: &str, limit: usize) -> Result<Vec<Track>> {
        let limit_str = limit.to_string();
        let stdout = self.run(&["search", query, &limit_str])?;
        let tracks: Vec<Track> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(tracks)
    }

    pub async fn get_stream_info(&self, track_id: u64) -> Result<StreamInfo> {
        let id_str = track_id.to_string();
        let stdout = self.run(&["stream", &id_str, self.quality.as_api_str()])?;
        let resp: StreamResp = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        if let Some(e) = resp.error { return Err(anyhow!("{e}")); }
        Ok(StreamInfo {
            url:         resp.url.ok_or_else(|| anyhow!("Sin URL de stream"))?,
            codec:       resp.codec.unwrap_or_else(|| "flac".into()),
            bit_depth:   resp.bit_depth.unwrap_or(16),
            sample_rate: resp.sample_rate.unwrap_or(44100),
        })
    }

    pub async fn get_cover(&self, track_id: u64) -> Result<CoverInfo> {
        let id_str = track_id.to_string();
        let stdout = self.run(&["cover", &id_str])?;
        let resp: CoverResp = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        if let Some(e) = resp.error { return Err(anyhow!("{e}")); }
        Ok(CoverInfo {
            url: resp.url.unwrap_or_default(),
        })
    }
    pub async fn get_user_playlists(&self) -> Result<Vec<Playlist>> {
        let stdout = self.run(&["playlists"])?;
        let playlists: Vec<Playlist> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(playlists)
    }

    pub async fn get_playlist_tracks(&self, uuid: &str) -> Result<Vec<Track>> {
        let stdout = self.run(&["playlist_tracks", uuid])?;
        let tracks: Vec<Track> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(tracks)
    }

    pub async fn get_user_mixes(&self) -> Result<Vec<Mix>> {
        let stdout = self.run(&["mixes"])?;
        let mixes: Vec<Mix> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(mixes)
    }

    pub async fn get_mix_tracks(&self, mix_id: &str) -> Result<Vec<Track>> {
        let stdout = self.run(&["mix_tracks", mix_id])?;
        let tracks: Vec<Track> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(tracks)
    }

    pub async fn get_track_radio(&self, track_id: u64) -> Result<Vec<Track>> {
        let stdout = self.run(&["radio", &track_id.to_string()])?;
        let tracks: Vec<Track> = serde_json::from_str(&stdout)
            .map_err(|e| anyhow!("JSON error: {e}\noutput: {stdout}"))?;
        Ok(tracks)
    }
}